### DISCUSSION TITLE
2.0 Migration Problem..

### DISCUSSION THREAD

**Discussion by Patrick386** [firebase/flutterfire#6043] [Q&A] [answered]
I am working on a 2.0 migration.
I used the Riverpod StreamProvider, and I have a problem.

`final CollectionReference<ServiceProductModel> serviceProductCollectionRef =
    firebaseFirestore
        .collection(SERVICE_PRODUCT_COLLECTION)
        .withConverter<ServiceProductModel>(
            fromFirestore:
                (DocumentSnapshot<Map<String, dynamic>> snapshot, _) =>
                    ServiceProductModel.formJson(snapshot.data()!),
            toFirestore: (ServiceProductModel model, _) => model.toJson());`

```
final AutoDisposeStreamProvider serviceProductStreamProvider =
    AutoDisposeStreamProvider((AutoDisposeProviderReference ref) {
  return serviceProductCollectionRef.snapshots(); //.map((QuerySnapshot<ServiceProductModel> snapshot) => snapshot.docs); //.map((event) => event.docs);
});
```

```
 Widget build(BuildContext context) {
     AsyncValue _svrProdStream =
        useProvider(serviceProductStreamProvider); 

return _svrProdStream.when(
      **data: (serviceProduct) {** <<-- **I don't know what type it is, so it makes me confusing.**
       .....
      }
```

1. Is it possible to get'List<ServiceProductModel>' via stream?
2. I need an example for StreamProvider.
Thank you.

**Comment by stargazing-dino** [comment, marked as answer]
When writing code snippets, make sure they're legible for others to better help you.

Please use [highlighting](https://docs.github.com/en/github/writing-on-github/creating-and-highlighting-code-blocks) and proper formatting. Darts language code is `dart`.

As for a way you can create a stream of product docs (if this is what you want) you can creating a stream of the snapshots and then consume that stream in another provider that exposes just the docs (This helps you separate concerns, the mapping of docs from the source stream).

```dart
final serviceProductSnapshotsProvider =
    StreamProvider.autoDispose<QuerySnapshot<ServiceProduct>>(
  (ref) => serviceProductCollectionRef.snapshots(),
);

final serviceProductDocsProvider =
    Provider.autoDispose<AsyncValue<List<DocumentSnapshot<ServiceProduct>>>>(
  (ref) {
    return ref.watch(serviceProductSnapshotsProvider).whenData((snapshot) {
      return snapshot.docs;
    });
  },
);
```

Or, if you prefer to just get the docs, you can do


```dart
final serviceProductDocsProvider =
    StreamProvider.autoDispose<List<QueryDocumentSnapshot<ServiceProduct>>>(
  (ref) => serviceProductCollectionRef.snapshots().map((snapshot) => snapshot.docs),
);
```

**Reply by Patrick386** [reply]
```
final serviceProductDocsProvider =
    StreamProvider.autoDispose<List<QueryDocumentSnapshot<ServiceProduct>>>(
  (ref) => serviceProductCollectionRef.snapshots().map((snapshot) => snapshot.docs),
);
```

**The length of docs and the documentation could not be read.
How should I read it?**
It's rather confusing as the Type isn't clear.

**Error**
Another exception was thrown: NoSuchMethodError: 'name'

```
Widget build(BuildContext context) {
  
    AsyncValue _svrProdDocs = useProvider(serviceProductDocsProvider);

    return _svrProdDocs.when(
      loading: ...
      error:  ...
      data: (prodDoc) {  //<< dynamic prodDoc ?
   
      **How do you read the data here?**
      print('serviceProduct: ${prodDoc[0].name}');
    
...
}
}
```

**Reply by stargazing-dino** [reply]
The length of docs can be read by `snapshot.size` if i remember or you can just get the call len on the returned list value of `serviceProductDocsProvider `.

In your code you override the type of _svrProdDocs as `AsyncValue`

```dart
// Now becomes AsyncValue<dynamic>
AsyncValue _svrProdDocs = useProvider(serviceProductDocsProvider);
```

Therefore you lose type information. You can either use `final` or type it fully as what it is

```dart
final _svrProdDocs = useProvider(serviceProductDocsProvider);
// or
AsyncValue<List<QueryDocumentSnapshot<ServiceProduct>>> _svrProdDocs = useProvider(serviceProductDocsProvider);
```
When you do that, `prodDoc` will no longer be dynamic

To get that data, you can just call `.data()` on the snapshot

```dart
final ServiceProduct product = prodDoc.data();
```

If all you wanted was your list of `ServiceProduct`s from the start, you can just change your provider like this

```dart
final serviceProductDocsProvider =
    StreamProvider.autoDispose<List<ServiceProduct>>(
  (ref) => serviceProductCollectionRef.snapshots().map((snapshot) => snapshot.docs.map((doc) => doc.data())),
);
```

**Reply by Patrick386** [reply]
I fully understood the method presented, but did not get the desired result. I've tried several methods, but I can't.
All packages prior to migration were up to date and worked fine.
I followed the migration documentation, but I can't read the data from the server.
I tested it again using the sample StreamBuilder from the documentation. 
There is an error in the sample and I cannot get the desired result again.

```
final AutoDisposeStreamProvider serviceProductDocsProvider =
    StreamProvider.autoDispose<List<ServiceProductModel>>(
        (AutoDisposeProviderReference ref) => serviceProductCollectionRef
            .snapshots()
            .map((QuerySnapshot<ServiceProductModel> snapshot) => snapshot.docs
                .map((QueryDocumentSnapshot<ServiceProductModel> doc) =>
                    doc.data())
                .toList()));
```

```
══╡ EXCEPTION CAUGHT BY WIDGETS LIBRARY ╞═══════════════════════════════════════════════════════════
The following assertion was thrown building Text("Error Expected a value of type 'String', but got
one of type 'Null'", dependencies: [DefaultTextStyle, MediaQuery]):

```


cloud_firestore: ^ 2.0.0
flutter_hooks: ^ 0.16.0
hooks_riverpod: ^ 0.14.0 + 3
freezed_annotation: ^ 0.14.1
json_annotation: ^ 4.0.1

**Reply by Patrick386** [reply]
It has been determined that the word for an unused field in the model is incorrect. I found that what was ignored before 2.0 works very strict. I went through all of the sources one by one and I ended up with a 2.0 migration success. 
Thank you.
